src/main/java/com/google/devtools/build/remote/client/ActionGrouping.java

```diff
@@ -21,46 +26,181 @@
 
   @VisibleForTesting static final String actionString = "Entries for action with hash '%s'\n";
 
-  // Key: actionId; Value: a set of associated log entries.
-  Map<String, Multiset<LogEntry>> actionMap = new HashMap<>();
-  int numSkipped = 0;
+  @VisibleForTesting
+  static class ActionDetails {
+    Multiset<LogEntry> log;
```

Comment:
All of the state variables can now be marked final, and  set as arguments of the private constructor called by the builder. Ditto in other classes that have a builder.

Reply:
Done